build: imgtool integration for MCUboot

mbolivar-nordic opened the pull request:
In https://github.com/zephyrproject-rtos/zephyr/issues/22562#issuecomment-654421847 and following comments, we discuss how to make it easier to build zephyr applications to be chain-loaded by MCUboot. In short, we agreed on auto-generating signed binaries, provided sufficient information about how to generate them is available in .config.

This PR implements that.

If accepted, I will write a follow-up PR that teaches `west flash` to use the signed binaries by default. This will probably involve finally deprecating the `--elf-file`, `--hex-file`, and `--bin-file` runner options and just loading the default locations from `build/zephyr/.config`, with enough wiggle room to override those defaults when MCUboot is in use.

We can leave an escape hatch for user-supplied files using the `--file` option we've discussed adding support for.

The first four patches are just cleanups I noticed along the way and prep work. (@finikorg, I hope you will ACK the rimage changes to support `west sign --quiet`).

The last patch is the main one.

How I am testing, for reference:

```
west build -b nrf52840dk_nrf52840 -s bootloader/mcuboot/boot/zephyr -d build-mcuboot
west flash -d build-mcuboot

west build -b nrf52840dk_nrf52840 -s zephyr/samples/hello_world -- \
   -DCONFIG_BOOTLOADER_MCUBOOT=y \
   -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"/home/mbolivar/zp/bootloader/mcuboot/root-rsa-2048.pem\"

# My board's default flash runner supports --hex-file.
west flash --hex-file build/zephyr/zephyr.signed.hex
```

mbolivar-nordic committed:
build: auto-generate signed mcuboot binaries

Add new Kconfig options which depend on CONFIG_BOOTLOADER_MCUBOOT:

- CONFIG_MCUBOOT_SIGNATURE_KEY_FILE: the path to the key pair which
  should be used to sign the image, in PEM format. This is sent to
  imgtool via 'west sign' when set to produce zephyr.signed.bin and
  zephyr.signed.hex files as needed.

- CONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS: additional arguments to pass to
  imgtool.

- CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE: also generate
  zephyr.signed.confirmed.{bin,hex}

Add build system support for these options.

This makes a separate 'west sign' step unnecessary when using MCUboot,
if the application image is properly configured with the location of
the key file.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
cmake: add IMGTOOL optional dependency

This will be used in subsequent patches to integrate imgtool into the
build when the user has set CONFIG_BOOTLOADER_MCUBOOT=y.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
build: update stale comment in boilerplate.cmake

The zephyr_property_target use cases have grown.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: west sign: add --quiet option

This suppresses stdout on normal runs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@JordanYates please review. (I can't add you as a reviewer but I would like your feedback.) Thanks.

JordanYates commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,107 @@
+  # explicitly, in case those defaults change.
+
+  if(CONFIG_BUILD_OUTPUT_BIN)
+    set(bin_args --bin --sbin ${ZEPHYR_BINARY_DIR}/zephyr.signed.bin)
```
Does it makes sense for the output binary to be ` ${ZEPHYR_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.signed.bin`?
Default behaviour would still be `zephyr.bin` and `zephyr.signed.bin`, but `custom_app_name.bin` would generate `custom_app_name.signed.bin`.
Same reasoning would apply to the hex output.

henrikbrixandersen replied:
I agree.

mbolivar-nordic replied:
> Does it makes sense for the output binary to be ${ZEPHYR_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.signed.bin?
> Same reasoning would apply to the hex output.

Agreed.

henrikbrixandersen commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,107 @@
+    # this is the documented behavior.
+    return()
+  elseif(NOT EXISTS "${CONFIG_MCUBOOT_SIGNATURE_KEY_FILE}")
+    message(WARNING "Not signing images for MCUboot: CONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"${CONFIG_MCUBOOT_SIGNATURE_KEY_FILE}\" does not exist.")
```
Should this be an error instead?

henrikbrixandersen commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,107 @@
+    message(WARNING "Not signing images for MCUboot: CONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"${CONFIG_MCUBOOT_SIGNATURE_KEY_FILE}\" does not exist.")
+    return()
+  elseif(NOT WEST)
+    message(WARNING "Not signing images for MCUboot: west not found.")
```
Should this be an error instead?

mbolivar-nordic replied:
> Should this be an error instead?

@henrikbrixandersen I think that's reasonable. The user has to take explicit action to set MCUBOOT_SIGNATURE_KEY_FILE to a non-default value so the intent is clear here.

henrikbrixandersen commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,107 @@
+  # Even though west is installed, imgtool might not be if the user has
+  # a custom manifest which doesn't include MCUboot.
+  #
+  # Therefore, go with an explicitly installed imgtool first, falling
+  # back on mcuboot/scripts/imgtool.py.
```
Would it make sense to turn this logic upside down? If mcuboot is part of the manifest, the version of imgtool known to work with that revision of mcuboot is the one shipped with it.

mbolivar-nordic replied:
> Would it make sense to turn this logic upside down? If mcuboot is part of the manifest, the version of imgtool known to work with that revision of mcuboot is the one shipped with it.

@henrikbrixandersen My thinking here is that the user has to install imgtool with pip explicitly, because it's not part of any requirements.txt file. So the intent if it's present seems to be to use that one; doing it this way also avoids a subprocess on Windows, where they're more expensive.

If you think this is important, I will reverse the order.

henrikbrixandersen replied:
No, that's a good point. I didn't think of that. Let's keep the order as-is.

henrikbrixandersen commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,107 @@
+
+  # No imgtool, no signed binaries.
+  if(NOT DEFINED imgtool_path)
+    message(WARNING "Not signing images with imgtool: can't find imgtool. To fix, install imgtool with pip3 and make sure it's in PATH, or add the mcuboot repository to the west manifest")
```
Again, should this be an error instead?

henrikbrixandersen reviewed (changes requested):
@mbolivar-nordic Thank you for working on this. It will greatly reduce the complexity of building signed binaries for most downstream users.

I wonder if the various warnings emitted if west or imgtool.py isn't found although the configuration calls for signing should be errors instead?

henrikbrixandersen commented:
In our downstream, we always generate both a pre-confirmed and a non-confirmed image when signing. This allows us to use the exact same .elf file for both production flash programming (where we use the pre-confirmed image and merge it with the mcuboot image) and OTA updates (where we use the non-confirmed image and confirm the image externally after upgrade).

Do you think this use-case is generic enough to be supported by the upstream build system? Perhaps with a Kconfig option for providing a pre-confirmed image as well?

mbolivar-nordic commented:
> Do you think this use-case is generic enough to be supported by the upstream build system? Perhaps with a Kconfig option for providing a pre-confirmed image as well?

Sure, that seems reasonable. Can you suggest some file names?

For example, should zephyr.signed.hex be unconfirmed, and zephyr.signed-confirmed.hex be confirmed?

mbolivar-nordic commented:
I pushed a v2 which changes warnings to errors and respects CONFIG_KERNEL_BIN_NAME.

Open review comments:

- creating confirmed images too, potentially under another Kconfig option (what to name the files in this case?)
- whether to prefer the imgtool that comes with mcuboot or the one installed with pip

henrikbrixandersen commented:
> Sure, that seems reasonable. Can you suggest some file names?
> 
> For example, should zephyr.signed.hex be unconfirmed, and zephyr.signed-confirmed.hex be confirmed?

@mbolivar-nordic Yes, that sounds like a reasonable naming scheme to me (or zephyr.signed.confirmed.hex?).

mbolivar-nordic commented:
Pushed a v3:

- Added CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE
- Documentation updates to Kconfig options
- Miscellaneous cleanups

No open review comments.

henrikbrixandersen reviewed (approved):
Thanks, @mbolivar-nordic. Looks good! Only thing missing, I can think of, is a test of this within the tree - but since the path to the key file must be absolute, it's probably hard to get it to work in CI.

mbolivar-nordic commented:
> Only thing missing, I can think of, is a test of this within the tree - but since the path to the key file must be absolute, it's probably hard to get it to work in CI.

That's not the issue as much as me having no idea how to write the test.

It's easy to write a test to test C or Python behavior, but I don't know of any tests that verify the build artifacts themselves.

I propose that we test this is from the `west flash` test cases once that tool understands how to react to these Kconfig options. Sound reasonable?

henrikbrixandersen commented:
> I propose that we test this from the `west flash` test cases once that tool understands how to react to these Kconfig options. Sound reasonable?

Yup.

mbolivar-nordic commented:
You know, since this feature depends on west, we could always make relative paths relative to the workspace root. I won't do that for this PR, but I'll give it a shot when modifying `west flash` to use this.

JordanYates commented:
Looks good to me

mbolivar-nordic commented:
Just a rebase.

mbolivar-nordic commented:
I added DNM because I would like to see https://github.com/zephyrproject-rtos/zephyr/pull/27740 merged first, so I can rebase on top without the extra shell out to west list.

tejlmand commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,130 @@
+    #
+    # Therefore, the build system doesn't regard it as a module, and
+    # doesn't define ZEPHYR_MCUBOOT_MODULE_DIR.
+    execute_process(COMMAND ${WEST} list -f "{posixpath}" mcuboot
```
I know this will change now, but remember, as per: https://github.com/zephyrproject-rtos/zephyr/pull/26060, we should avoid calling `west` directly, but instead through same Python as rest of build system.

tejlmand commented on cmake/mcuboot.cmake:
```diff
@@ -0,0 +1,130 @@
+  endif()
+
+  # Basic 'west sign' command and output format independent arguments.
+  set(west_sign ${WEST} sign --quiet --tool imgtool
```
This should be using same python as rest of build system instead.
See https://github.com/zephyrproject-rtos/zephyr/pull/26060

mbolivar-nordic replied:
OK, I will need to think about how to do that in a convenient way.

I don't want to release 0.7.4 and would rather just get 0.8 out the door with this. We can add a hack for 0.7.x that breaks abstraction barriers. Cutting a new west release requires enough testing that I don't want to do it unless I have to :smile:

tejlmand reviewed (changes requested):
We should use same python as rest of build system, when calling `west`.
See https://github.com/zephyrproject-rtos/zephyr/pull/26060.

Would `python -m west sign ...` be of help, or is that too early, cause it would force everyone to use a newer `west`

mbolivar-nordic commented:
Open issues:

- now blocked on #27740 
- need to run west using the same python as everything else
- I've tested making relative paths relative to `west topdir` and it's working fine, so I'll push that into this PR since it was a trivial change and we're blocked now

I have a draft PR open for making `west flash` use the signed artifacts as well: https://github.com/zephyrproject-rtos/zephyr/pull/27818. It's not ready to merge, but I thought I'd mention it since this is back to being blocked in case people are curious.

mbolivar-nordic committed:
cmake: west detection fixes and improvements

Starting with west 0.8, we will have a stable way to run west as a
python module. In west 0.7, we have an undocumented way.

Handle these two cases to override the WEST variable used by the build
system to run west inside the same python interpreter as the rest of
the build system. This makes it intuitive to shell out to west from
elsewhere in the build, without making people reimplement existing
commands in the west API when it is not necessary.

Do so for the 'west topdir' invocation in west.cmake as an example.

The check that looks for whether PYTHON_EXECUTABLE and WEST_PYTHON
match is also not working properly when symbolic links are in use.
Fix it by using REALPATH instead of ABSOLUTE in the
get_filename_component() calls.

Finally, print the error on a failed import that we expect to work.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Pushed a v4:

- rebased onto #27740 plus one DNM fix that ought to go into that PR (@tejlmand : I requested changes there)
- added a new patch, `cmake: west detection fixes and improvements`, that makes it so WEST runs west under the same python interpreter as the rest of the system (@tejlmand this addresses your open comment, please revisit)
- relative paths are now relative to `west topdir`

Open issues:

- blocked on #27740

tejlmand commented on cmake/west.cmake:
```diff
@@ -1,60 +1,91 @@
-  # even after output is one line.
-  message(STATUS "Found west: ${WEST} (found suitable version \"${west_version}\", minimum required is \"${MIN_WEST_VERSION}\")")
+  # Set WEST to a COMMAND prefix as if it were a find_program() result.
+  if(${west_version} VERSION_GREATER_EQUAL 0.8)
+    # The 'west' package became runnable as a module during 0.8
+    # development.
+    set(WEST_MODULE west)
+  else()
+    # In west 0.7.x, it's possible to run west.app.main. Simplify this
+    # once 0.8.0 is the minimum supported version.
+    set(WEST_MODULE west.app.main)
+  endif()
```
How about only special handling of `0.7.x` like:

```suggestion
set(WEST_MODULE west)
if(${west_version} VERSION_LESS 0.8)
  # In west 0.7.x, west is not directly available as a module so we need to specify full namespace to main.
  string(APPEND WEST_MODULE ".app.main")
endif()
```

Another thing, a clever man once said:
> Same thing here; no need to expose the CLI implementation details. We can just use the west API directly.

(sorry, couldn't resist :laughing: )

mbolivar-nordic replied:
> How about only special handling of `0.7.x` like:

Sure, that's fine.

Done and done.



> Another thing, a clever man once said:

Heh

tejlmand reviewed (commented):
Looks good. Just a small suggestion.

mbolivar-nordic commented:
Pushed v5:

- rebased on master now that #27740 is merged
- applied @tejlmand's suggested change to the new patch from v4

No open issues. @tejlmand please revisit and approve if satisfied.

tejlmand reviewed (approved):
lgtm

marc-hb commented on scripts/west_commands/sign.py:
```diff
@@ -305,14 +312,15 @@ def get_cfg(command, bcfg, item):
         dts = b / 'zephyr' / 'zephyr.dts'
-        log.dbg('DTS file:', dts, level=log.VERBOSE_VERY)
+        if not quiet:
+            log.dbg('DTS file:', dts, level=log.VERBOSE_VERY)
```
I really doubt this works. I also tried to https://pylint.pycqa.org/en/latest/user_guide/messages/warning/logging-not-lazy.html  somewhere else but it did not work; this is not the standard python logger unfortunately.

mbolivar-nordic replied:
> I really doubt this works.

This PR was merged over 2 years ago; please be more specific or file a bug :)

marc-hb replied:
Sorry for the noise, I didn't expect `log.dbg()` to work like this and notably to take a `level` argument while the other... levels (!) don't take any. It's not symmetric and intuitive.

What I should definitely have expected from you is for all that to be thoroughly documented:
https://docs.zephyrproject.org/1.14.0/guides/west/west-apis.html#module-west.log

My bad!
